src/presence/HeartbeatComponent.js

```diff
@@ -0,0 +1,63 @@
+// @flow strict-local
+import { PureComponent } from 'react';
+import { AppState } from 'react-native';
+import type { Auth, Dispatch } from '../types';
+
+import { connect } from '../react-redux';
+import { tryGetAuth } from '../account/accountsSelectors';
+import { reportPresence } from '../actions';
+import Heartbeat from './heartbeat';
+
+type Props = $ReadOnly<{|
+  dispatch: Dispatch,
+  auth: Auth | void,
+|}>;
+
+/**
+ * Component providing a recurrent `presence` signal.
+ */
+// Note that "PureComponent" is of questionable veracity: this component's
+// entire purpose is to emit network calls for their observable side effects.
+// However, it is at least true that there is never a need to call `render()` if
+// the props haven't changed.
+class PresenceHeartbeat extends PureComponent<Props> {
+  /** Callback for Heartbeat object. */
+  onHeartbeat = (state: boolean) => {
+    // N.B.: If `auth` changes, we do not send out a final `false` presence
+    // status for the previous `auth`. It's the responsibility of our logout
+    // handler to determine whether that's necessary.
+    //
+    // (TODO: ensure that our logout handlers actually do that.)
+    if (this.props.auth) {
+      this.props.dispatch(reportPresence(state));
+    }
+  };
+
+  heartbeat: Heartbeat = new Heartbeat(this.onHeartbeat, 1000 * 60);
+
+  componentDidMount() {
+    this.onStateChange();
+    AppState.addEventListener('change', this.onStateChange);
+  }
+
+  componentWillUnmount() {
+    AppState.removeEventListener('change', this.onStateChange);
```

Comment:
What's the thinking behind using this vs. the `isActive` we maintain in Redux? (With the listener in `AppEventHandlers`.)

Reply:
Mostly the fact that `isActive` isn't really application-_controlled_ state. There are a lot of things that may happen in between that handler and a call to `render()` here.

(For example: if we get rehydrated with the wrong `isActive` state, what happens? If that's not possible, why not?)

Reply:
> (For example: if we get rehydrated with the wrong `isActive` state, what happens? If that's not possible, why not?)

That's not possible because `isActive` is more fully `state.session.isActive`, and the `session` subtree is one we don't persist. See `src/boot/store.js`:
```
/**
 * Properties on the global store which we explicitly choose not to persist.
 *
 * All properties on the global store should appear either here or in the
 * lists of properties we do persist, below.
 */
// prettier-ignore
export const discardKeys: Array<$Keys<GlobalState>> = [
  'alertWords', 'caughtUp', 'fetching', 'loading',
  'nav', 'presence', 'session', 'topics', 'typing', 'userStatus',
];
```


> Mostly the fact that `isActive` isn't really application-_controlled_ state.

Hmm, can you unpack what you mean by "controlled"? I might not understand what you mean. Or does the bit above about persistence affect this?

This state is completely determined by our code, specifically `sessionReducer.js`.

Reply:
Sorry – the `.isActive` field itself is certainly owned and controlled by our app, but its value is a representation and a cache of underlying state that very much is not.

Notably, if the app is *not* active, it affects the cached value's staleness period in ways that are nontrivial to reason about – here, based on the decisions of two different schedulers, one of which is OS-dependent. To me that's an indication that this data probably shouldn't be in the store at all.

In this particular use case... honestly, it's probably not that important if we don't get the notice promptly. If there's a reason it *should* go through Redux, I've got a short patch I can squash into the relevant commit.

(On the other hand, I'd argue that the lone existing use of `.isActive`, over in `src/events/eventMiddleware.js`, is a bug! Or at least is closely involved in one. But that's not directly relevant here; I'll elaborate somewhere else.)